### PR TITLE
[android][constants] Rewrite android code to Kotlin

### PR DESCRIPTION
**Pull request by kkafar**
# Why

This task was done as a part of migrating expo modules to Kotlin project.

# How

The code had been initially converted from Java to Kotlin with help of Android Studio's conversion tool. Then it was manually adjusted / improved.

# Test Plan

- [x] Test suite
- [x] NCL

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.io and README.md).
- [x] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).

### CODE REVIEW

File: packages/expo-constants/android/src/main/java/expo/modules/constants/ConstantsService.kt [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+          stream ->
+          return IOUtils.toString(stream, StandardCharsets.UTF_8)
+        }
+      } catch (e: FileNotFoundException) {
+        // do nothing, expected in managed apps
+      } catch (e: Exception) {
+        Log.e(TAG, "Error reading embedded app config", e)
+      }
+      return null
+    }
```
Comment (kkafar) [inline packages/expo-constants/android/src/main/java/expo/modules/constants/ConstantsService.kt:113]: I was not able to find better alternative for java's [`try-with-resources`](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html) syntax than `use`. Suggestions here are especially welcome.

**Review by kkafar — COMMENTED** [review]
Notes:

**Review by bbarthec — APPROVED** [review]
It looks okayish to me 👍

File: packages/expo-constants/android/src/main/java/expo/modules/constants/ConstantsModule.kt [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+import expo.modules.core.ExportedModule
+import expo.modules.core.ModuleRegistry
+import expo.modules.core.ModuleRegistryDelegate
+import expo.modules.core.Promise
+import expo.modules.interfaces.constants.ConstantsInterface
+import expo.modules.core.interfaces.ExpoMethod
+
+class ConstantsModule(
+  context: Context,
+  private val moduleRegistryDelegate: ModuleRegistryDelegate = ModuleRegistryDelegate()
```
[the remark this answers is not in the document]
Comment (barthap) [reply packages/expo-constants/android/src/main/java/expo/modules/constants/ConstantsModule.kt:15]: This is in the constructor in other modules too (see `expo-image-picker`, `expo-speech`) so he's just keeping convention. When we finally kotlinize `expo-modules-core`, we can move that into `ExportedModule` class and make `protected`

**Review by barthap — COMMENTED** [review]
LGTM in overall,

I had some concerns about `ExponentInstallationService` and its other usages but this instance is used only inside `expo-constants`

I found out that it's duplicated in:
- [expoview](https://github.com/expo/expo/blob/master/android/expoview/src/main/java/host/exp/exponent/storage/ExponentInstallationId.kt) (already kotlinized)
- [expo-notifications](https://github.com/expo/expo/blob/1c9ac2673344cbb3aefa0c67069937cff20ab7a3/packages/expo-notifications/android/src/main/java/expo/modules/notifications/serverregistration/InstallationId.java)

**Review by barthap — APPROVED** [review]

**Pull request merged**